### PR TITLE
CATO-4894  Modified CP33 to allow positive and negative numbers inclu…

### PR DESCRIPTION
**Pull request by rutpatelhmrc**
CATO-4894  Modified CP33 to allow positive and negative numbers including a new method & trait that supports this logic 

Large file changes due to year change

### CODE REVIEW

**Comment by kastoestoramadus** [comment]
You've attached year update to this PR. It makes the ticket changes unreadable.
Create second PR just for 2017->2018 update.

File: src/main/scala/uk/gov/hmrc/ct/box/Types.scala [hunk 1]
```diff
@@ -244,6 +244,14 @@ trait CtOptionalDate extends OptionalCtValue[LocalDate] {
 
 }
 
+trait IsNegativeOrPositive {
+
+  self: CtInteger =>
+
+  require(value >= 0 || value <= 0, "This box must be either positive or negative")
```
Comment (kastoestoramadus) [inline src/main/scala/uk/gov/hmrc/ct/box/Types.scala]: does exist a number not passing this condition?

File: src/main/scala/uk/gov/hmrc/ct/box/Types.scala [hunk 2]
```diff
@@ -244,6 +244,14 @@ trait CtOptionalDate extends OptionalCtValue[LocalDate] {
 
 }
 
+trait IsNegativeOrPositive {
```
[the remark this answers is not in the document]
Comment (kastoestoramadus) [reply src/main/scala/uk/gov/hmrc/ct/box/Types.scala]: Intelij haven't found any usages of this trait. Why you've added it?

File: src/main/scala/uk/gov/hmrc/ct/box/ValidatableBox.scala [hunk 3]
```diff
@@ -176,6 +176,12 @@ trait ValidatableBox[T <: BoxRetriever] extends Validators {
     }
   }
 
+  protected  def isPositiveOrNegative(box: OptionalIntIdBox)(): Set[CtValidation] = {
```
Comment (kastoestoramadus) [inline src/main/scala/uk/gov/hmrc/ct/box/ValidatableBox.scala]: naming issue, out of convention for validations

File: src/test/scala/uk/gov/hmrc/ct/BoxValidationFixture.scala [hunk 4]
```diff
@@ -28,6 +28,21 @@ trait BoxValidationFixture[T <: ComputationsBoxRetriever] extends WordSpec with
   //This can be overridden if mock box retriever calls need to be made
   def setUpMocks(): Unit = Unit
 
+  def testIsPositiveOrNegative(boxId: String, builder: Option[Int] => ValidatableBox[T]) = {
```
Comment (kastoestoramadus) [inline src/test/scala/uk/gov/hmrc/ct/BoxValidationFixture.scala]: Type system already checks if the value is an Integer. 
the validator may have a name like: noConstrainsValidation or not exist at all if it is easy to do.

File: src/main/scala/uk/gov/hmrc/ct/box/ValidatableBox.scala [hunk 5]
```diff
@@ -176,6 +176,12 @@ trait ValidatableBox[T <: BoxRetriever] extends Validators {
...
+    box.value match {
+      case Some(x) => Set()
+    }
```
Comment (liquidarmour) [inline src/main/scala/uk/gov/hmrc/ct/box/ValidatableBox.scala]: match error.

**Comment by liquidarmour** [comment]
I'm glad we closed this.

To be honest all you have to do is remove the validate function from CP33. Nothing else is required unless there is a range we want to enforce.

**Pull request closed without merging**